### PR TITLE
Support case default (towards #354)

### PR DESCRIPTION
**Pull request by arporter**
Adds support to the PSyIR for SELECT blocks containing a CASE DEFAULT clause.

### CODE REVIEW

File: src/psyclone/psyGen.py [hunk 1]
```diff
@@ -5575,9 +5578,14 @@ def _case_construct_handler(self, node, parent):
...
-        # Search for all the CASE clauses in the Case_Construct
+        # Search for all the CASE clauses in the Case_Construct. We do this
+        # because the fp2 parse tree has a flat structure at this point with
+        # the clauses being siblings of the contents of the clauses. The
+        # final index in this list will hold the position of the end-select
+        # statement.
         clause_indices = []
         selector = None
+        default_clause_idx = -1
         for idx, child in enumerate(node.content):
```
Comment (sergisiso) [inline src/psyclone/psyGen.py]: Update `# Case Default and value Ranges not supported yet, if found` comment

File: src/psyclone/psyGen.py [hunk 2]
```diff
@@ -5575,9 +5578,14 @@ def _case_construct_handler(self, node, parent):
...
 
-        # Search for all the CASE clauses in the Case_Construct
+        # Search for all the CASE clauses in the Case_Construct. We do this
+        # because the fp2 parse tree has a flat structure at this point with
+        # the clauses being siblings of the contents of the clauses. The
+        # final index in this list will hold the position of the end-select
+        # statement.
         clause_indices = []
         selector = None
+        default_clause_idx = -1
```
Comment (sergisiso) [inline src/psyclone/psyGen.py]: Maybe `default_clause_idx = None` and then ` if default_clause_idx:` is more pythonic?

File: src/psyclone/tests/psyGen_test.py [hunk 3]
```diff
@@ -4158,7 +4158,44 @@ def test_fparser2astprocessor_handling_Case_construct(f2008_parser):
...
+def test_fp2astproc_case_default(f2008_parser):
+    ''' Check that the fparser2ASTProcessor handles SELECT blocks with
+    a default clause. '''
+    from fparser.common.readfortran import FortranStringReader
+    from fparser.two.Fortran2003 import Execution_Part, Assignment_Stmt
+    case_clauses = ["CASE default\nbranch3 = 1\nbranch3 = branch3 * 2\n",
+                    "CASE (label1)\nbranch1 = 1\n",
+                    "CASE (label2)\nbranch2 = 1\n"]
+    # Loop over the 3 possible locations for the 'default' clause
+    for idx1, idx2, idx3 in [(0,1,2), (1,0,2), (1,2,0)]:
```
Comment (sergisiso) [inline src/psyclone/tests/psyGen_test.py]: Nice test :)

Comment (sergisiso) [reply src/psyclone/tests/psyGen_test.py]: But pylint complains about the missing whitespace separation after comas.

File: src/psyclone/psyGen.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/psyGen.py]: Oh yes, well spotted.

File: src/psyclone/psyGen.py (hunk 2 above)
Comment (arporter) [reply src/psyclone/psyGen.py]: Yes, I agree. Done.

File: src/psyclone/tests/psyGen_test.py (hunk 3 above)
Comment (arporter) [reply src/psyclone/tests/psyGen_test.py]: Sigh :-)

**Comment by arporter** [comment]
Assuming Travis is happy then this is ready for another review...

**Pull request merged**